### DISCUSSION TITLE
Print the matrix in the MOOSE module

### DISCUSSION THREAD

**Discussion by jinca** [idaholab/moose#17089] [Q&A Getting Started] [answered]
Hello guys,
May I ask how to print the type of the matrix used in the kernel?
Is OK to use the following?:
For NEWTON, you might use  petsc_options_iname = -ksp_view_mat 
For PJFNK, you might use  petsc_options_iname = -ksp_view_pmat 
Thanks in advance,
Julita

**Comment by aeslaughter** [comment]
@fdkong Can you help out here?

**Comment by fdkong** [comment, marked as answer]
> May I ask how to print the type of the matrix used in the kernel?

The matrix type is `MPIAIJ` (https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Mat/MATMPIAIJ.html) for parallel simulations, and `SeqAIJ` (https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Mat/MATSEQAIJ.html) for sequential simulations.  We do not explore other matrix types.

>For NEWTON, you might use petsc_options_iname = -ksp_view_mat
>For PJFNK, you might use petsc_options_iname = -ksp_view_pmat

Regardless, you could use  `-ksp_view_pmat` to print out matrix elements on the screen. But I want to point out this print is just for debugging. That being said, the option is often employed to print out a small matrix for checking if the code works correctly. 

If you want to dump a large matrix and use it in other codes, you need to use a binary file (-ksp_view_pmat binary). In general, I do NOT think it is a good idea to dump a large matrix unless you have a strong motivation. Instead, you should directly couple MOOSE to other codes to accomplish this.
